### ISSUE TITLE
Custom Image in SEO Image Source not Saving

### ISSUE THREAD

**Issue by artsigns** [nystudio107/craft-seomatic#416]
After migrating from Craft 2.x to 3.x and following the steps to set up SEOmatic and the new SEOmatic field, I am having a hard time getting Custom SEO Images to save correctly.

After saving the custom image, it just reverts to the Global SEO Image. 

I have tried re-saving each Content SEO section. Specifically going into General, Twitter, and Facebook sub-tabs a saving each.

I have tried removing the Custom Image from SEO Image Source and re-adding it, however that doesn't seem to help.

One thing I've noticed is when I double click on the image I can see under Focal Point, it is showing "Unable to find component class 'FocalPointField_FocalPoint'. However some of the Custom SEO Images are still working, even with this error.

I would not mind setting the SEO Image Source through Content SEO if there was a way to select the first image of a Matrix field, which is the image I normally use for the Custom Image.

Also, should I be deleting the old SEO Meta field?

**Comment by khalwat** [comment]
You can delete the old SEOmatic field if you like.

It sounds like some of the sources were coming from a Focal Point field, for which there is no Craft 3.x field.

You can have SEOmatic pull out that image from your matrix block if you want, but you'd need to write some Twig code:

https://github.com/nystudio107/craft-seomatic#twig-templating

**Comment by khalwat** [comment]
Does the above help @artsigns ?

**Comment by artsigns** [comment]
I deleted the old SEOmatic field but I am still unable to save a Custom Image under an Entry's SEO Image Source.

I also tried deleting the Custom Image, saving, and then re-adding it and saving again without success. 

The strange thing is, right after migrating from Craft 2.x to 3.x, I was able to successfully populate the Entry's Custom SEO Image by simply re-saving each entry. However, I only did so for the first few entries and that method no longer seems to work.

I also tried removing the custom image from the section's Content SEO. Which didn't change anything.
Then I tried changing the section's Custom Image, and I noticed it was just reverting back to the global image. So I deleted the global image which was the first time I saw the SEO Image change to nothing when testing in Browser. 

Without a global image set, the Custom Content SEO image for the channel still isn't showing. It seems like other than the Global SEO Image, the SEO image field either isn't letting me save new images to it or isn't falling back (or forward) correctly for me. Non of the Custom SEO images, neither by Content SEO for sections or set individually per entry are working (other than the ones I re-saved after the initial migration).

![Screen Shot 2019-06-11 at 1 06 43 PM](https://user-images.githubusercontent.com/23419938/59291888-d4ab2d00-8c49-11e9-8726-e41c89f51b25.jpg)

**Comment by artsigns** [comment]
I forgot to mention, I also deleted the old focal point field to ensure it wasn't interfering.

**Comment by khalwat** [comment]
So to be clear, the place where you are unable to save an image is in an SEO Settings field in an Entry?

**Comment by CreateSean** [comment]
I too am having this issue.

I updated Craft to 3.1.30 and SEOmatic to 3.2.8 this morning and then started writing a blog post. This blog post needs a custom SEO image rather than using the default hero image.

In the SEOmatic field I changed the dropdown to custom image and uploaded an image but it's not being reflected in the SEO as you can see from the screenshot.  - the tiny thumb is what I expect to be the SEO iamge, but what I get is the larger one on the right (Hero Image)
![seo-image1](https://user-images.githubusercontent.com/6899490/59451091-a7d95000-8dd8-11e9-8428-2bbc9c8774b0.jpg)

I can save the custom image but it is not being applied to the entry.

Additionally the twitter/facebook/sitemap tabs in that field are not changing when clicked. I noticed a console error which may help

```
_object-keys-internal.js:4 Uncaught TypeError: i(...) is not a function
    at Object.<anonymous> (_object-keys-internal.js:4)
    at c (bootstrap:79)
    at Object.<anonymous> (_object-keys.js:2)
    at c (bootstrap:79)
    at i (bootstrap:45)
    at bootstrap:152
    at bootstrap:152
(anonymous) @ _object-keys-internal.js:4
c @ bootstrap:79
(anonymous) @ _object-keys.js:2
c @ bootstrap:79
i @ bootstrap:45
(anonymous) @ bootstrap:152
(anonymous) @ bootstrap:152
6garnish.js:220 Double-instantiating a menu button on an element
garnish.js:220 Double-instantiating a menu button on an element
```

**Comment by artsigns** [comment]
Yes, when I save the SEO Image Source of the SEO Settings field in an entry, it saves but the SEO Preview Twitter and og:image stays as `NO IMAGE SET`.

I was under the impression that setting the SEO Image source under "Content SEO > Section > General" would by default also set an image for Twitter and Facebook Images. I now see that is not the case, and it seems you need to explicitly set an image for both Twitter and Facebook Images or go into each and set `Same As SEO Image`.

This led me to display the Twitter and Facebook Tabs in the entry SEO Settings, with the Twitter Image and Facebook OpenGraph Image fields.

Exposing these fields shows that they had no image set. I set the Twitter and Facebook Images manually and found that now the SEO Image is working.

Is it possible that the Twitter and Facebook Images somehow got lost or failed to set during the 2.x to 3.x migration?

Having the entry SEO Image Source be the default fallback image for both Twitter and Facebook Image sources may be a good solution.

**Comment by artsigns** [comment]
> I too am having this issue.
> 
> I updated Craft to 3.1.30 and SEOmatic to 3.2.8 this morning and then started writing a blog post. This blog post needs a custom SEO image rather than using the default hero image.
> 
> In the SEOmatic field I changed the dropdown to custom image and uploaded an image but it's not being reflected in the SEO as you can see from the screenshot. - the tiny thumb is what I expect to be the SEO iamge, but what I get is the larger one on the right (Hero Image)
> ![seo-image1](https://user-images.githubusercontent.com/6899490/59451091-a7d95000-8dd8-11e9-8428-2bbc9c8774b0.jpg)
> 
> I can save the custom image but it is not being applied to the entry.
> 
> Additionally the twitter/facebook/sitemap tabs in that field are not changing when clicked. I noticed a console error which may help
> 
> ```
> _object-keys-internal.js:4 Uncaught TypeError: i(...) is not a function
>     at Object.<anonymous> (_object-keys-internal.js:4)
>     at c (bootstrap:79)
>     at Object.<anonymous> (_object-keys.js:2)
>     at c (bootstrap:79)
>     at i (bootstrap:45)
>     at bootstrap:152
>     at bootstrap:152
> (anonymous) @ _object-keys-internal.js:4
> c @ bootstrap:79
> (anonymous) @ _object-keys.js:2
> c @ bootstrap:79
> i @ bootstrap:45
> (anonymous) @ bootstrap:152
> (anonymous) @ bootstrap:152
> 6garnish.js:220 Double-instantiating a menu button on an element
> garnish.js:220 Double-instantiating a menu button on an element
> ```

**Comment by artsigns** [comment]
@CreateSean Does it work when you save the Twitter and Facebook Images individually?

**Comment by CreateSean** [comment]
@artsigns I am unable to try that as the tabs don't work for me. Detailed above including console errors.

**Comment by khalwat** [comment]
So I'm unable to reproduce this on my local setup.

Can you two tell me, for your SEO Settings field... what things are enabled in the field settings? A screenshot would be fantastic.

Also the preview in question is the OpenGraph preview -- what do you have your OpenGraph image set to both in the SEO Settings field (assuming that's included there) and also in Content SEO?

For me, it's Same As SEO Image and the override via the field is working as expected.

**Comment by CreateSean** [comment]
Andrew have attached screenshots of both
![content-seo](https://user-images.githubusercontent.com/6899490/59461394-48d30580-8def-11e9-9399-008a353872a9.jpg)
![seo-field-settings](https://user-images.githubusercontent.com/6899490/59461395-48d30580-8def-11e9-92e2-47eea7f8dd51.jpg)

**Comment by khalwat** [comment]
Right @CreateSean so the Facebook Image is separate from the SEO Image. People wanted the ability to have a different image for Twitter than for Facebook, for instance.

So in your case, if you're changing just the SEO Image, that will not change the Facebook Image, because you have that set to pull from a field.

If you want it to change the SEO Image too, you'd need to change it to be "Same as SEO Image"

**Comment by artsigns** [comment]
Previously, I had the bare essentials for my SEO Settings field (Title, Description, Image). I just enabled the Twitter and Facebook tabs in order to gain access to their images for troubleshooting.

You are right, I see that if I set the Twitter and Facebook Image Source in their respective SEO Settings field Tabs to `Same as SEO Image`, then it begins to work as expected. However, the default in my case was Custom Image [Select Twitter / Facebook Image]. With the Twitter and Facebook tabs disabled by default it may be difficult for someone to understand why the image isn't working.

![Screen Shot 2019-06-13 at 3 19 17 PM](https://user-images.githubusercontent.com/23419938/59461119-a4e95a00-8dee-11e9-9b7f-a307c1d73c92.jpg)
![Screen Shot 2019-06-13 at 10 01 24 PM](https://user-images.githubusercontent.com/23419938/59478328-ebf34180-8e26-11e9-9ce2-e46b29e85fee.jpg)
